pkg/kubelet/cm/internal_container_lifecycle_test.go

```diff
@@ -0,0 +1,121 @@
+package cm
+
+import (
+	"github.com/golang/mock/gomock"
+	"k8s.io/api/core/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	kubefeatures "k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/memorymanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+	"testing"
+)
+
+func TestPreStartContainer(t *testing.T) {
+	// setup vars
+
+	var tests = []struct {
+		i internalContainerLifecycleImpl
+	}{
+		{
+			i: internalContainerLifecycleImpl{
+				cpuManager:      nil,
+				memoryManager:   nil,
+				topologyManager: nil,
+			},
+		},
+		{
+			i: internalContainerLifecycleImpl{
+				cpuManager:      cpumanager.NewFakeManager(),
+				memoryManager:   nil,
+				topologyManager: nil,
+			},
+		},
+		{
+			i: internalContainerLifecycleImpl{
+				cpuManager:      cpumanager.NewFakeManager(),
+				memoryManager:   memorymanager.NewFakeManager(),
+				topologyManager: nil,
+			},
+		},
+		{
+			i: internalContainerLifecycleImpl{
+				cpuManager:      cpumanager.NewFakeManager(),
+				memoryManager:   nil,
+				topologyManager: topologymanager.NewFakeManager(),
+			},
+		},
+		{
+			i: internalContainerLifecycleImpl{
+				cpuManager:      nil,
+				memoryManager:   memorymanager.NewFakeManager(),
+				topologyManager: nil,
+			},
+		},
+		{
+			i: internalContainerLifecycleImpl{
+				cpuManager:      nil,
+				memoryManager:   memorymanager.NewFakeManager(),
+				topologyManager: topologymanager.NewFakeManager(),
+			},
+		},
+		{
+			i: internalContainerLifecycleImpl{
+				cpuManager:      nil,
+				memoryManager:   nil,
+				topologyManager: topologymanager.NewFakeManager(),
+			},
+		},
+		{
+			i: internalContainerLifecycleImpl{
+				cpuManager:      cpumanager.NewFakeManager(),
+				memoryManager:   memorymanager.NewFakeManager(),
+				topologyManager: topologymanager.NewFakeManager(),
+			},
+		},
+	}
+
+	for _, test := range tests {
+		mockCtrl := gomock.NewController(t)
+		defer mockCtrl.Finish()
+
+		pod := &v1.Pod{}
+		container := &v1.Container{}
+		containerId := "test-id"
+		i := test.i
+
+		//("Running PreStart Container",i.PreStartContainer())
+		i.PreStartContainer(pod, container, containerId)
+
+		mockCPUManager := cpumanager.NewMockManager(mockCtrl)
+		mockMemoryManager := memorymanager.NewMockManager(mockCtrl)
+		mockTopologyManager := topologymanager.NewMockManager(mockCtrl)
+
+		if i.cpuManager != nil {
+			mockCPUManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(1)
+		} else {
+			mockCPUManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(0)
+		}
+
+		if i.memoryManager != nil {
+			mockMemoryManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(1)
+		} else {
+			mockMemoryManager.EXPECT().AddContainer(pod, container, containerId).Return(nil).Times(0)
+		}
+
+		if i.topologyManager != nil && utilfeature.DefaultFeatureGate.Enabled(kubefeatures.TopologyManager) {
```

Comment:
This case isn't quite right - We need to test with both the feature gate enabled and disabled explicitly in the test

Reply:
In the disabled case I expect to be called 0 times. I think the else case will cover it . Are you saying this as to make the code more readable ?

Reply:
for the disabled case, will be covered in the else case. Are you saying this so as to make is more readable ?

Reply:
@endocrimes Thanks for your inputs . I have made a few changes. I am still confused between the FakeManager and MockManager(which I added) should I extend the FakeManager or create a Mock Manager( as I have already done). Also I face issue "missing call" probably

Reply:
@endocrimes Thanks for your inputs . I have made a few changes. I am still confused between the FakeManager and MockManager(which I added) should I extend the FakeManager or create a Mock Manager( as I have already done).


 Also I face issue "missing call" probably